src/intrinsics/ecma262/Object.js

```diff
@@ -82,36 +83,55 @@ export default function(realm: Realm): NativeFunctionValue {
 
         let frm_was_partial = frm.isPartialObject();
         if (frm_was_partial) {
-          if (!frm.isSimpleObject()) {
-            // If this is not a simple object, it may have getters on it that can
+          if (!to.isSimpleObject() || !frm.isSimpleObject()) {
+            // If an object is not a simple object, it may have getters on it that can
             // mutate any state as a result. We don't yet support this.
             AbstractValue.reportIntrospectionError(nextSource);
             throw new FatalError();
           }
 
           to_must_be_partial = true;
-          // Make this temporally not partial
+          // Make this temporarily not partial
           // so that we can call frm.$OwnPropertyKeys below.
           frm.makeNotPartial();
         }
 
         if (to_must_be_partial) {
-          // Generate a residual Object.assign call that copies the
-          // partial properties that we don't know about.
-          AbstractValue.createTemporalFromBuildFunction(
-            realm,
-            ObjectValue,
-            [ObjectAssign, target, nextSource],
-            ([methodNode, targetNode, sourceNode]: Array<BabelNodeExpression>) => {
-              return t.callExpression(methodNode, [targetNode, sourceNode]);
+          if (to instanceof AbstractObjectValue && to.values.isTop() && to.isSimpleObject()) {
+            // We know to is well behaved, but we don't know any of its properties.
+            // Hence the fact that we also know nothing about the properties of frm does not cause a problem for us.
+            // Consequently, generate a residual Object.assign call that copies the properties that we don't know about.
```

Comment:
I don't think this is safe because we don't know anything about this object's identity. It might have the same identity as something else in our heap. Normally we don't allow mutations on an object that doesn't have known values in the domain:

https://github.com/facebook/prepack/blob/master/src/values/AbstractObjectValue.js#L482-L485

```js
var obj = {x:'a'}
var abstractObj = doSomethingThatYieldsAnTopValObjectFrom(obj);
Object.assign(abstractObj, {x:'b'});
result = obj.x;
```

In this scenario, Prepack doesn't know that `obj` could've been mutated.

Reply:
Good point. It would only be safe if to were also an intrinsic object and then only because of our overall assumption that intrinsic objects are never aliased. I'll tighten this up.

Reply:
That's an interesting point. Should these intrinsic objects be modeled as a single element (template) which is a partial object, instead of topVal? Why would you want to model an intrinsic object as topVal?

Reply:
You probably don't want an intrinsic with a values domain of Top.